File: leetcode/100-999/145. Binary Tree Postorder Traversal/solution_3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution 
{
    public:
        vector<int> postorderTraversal(TreeNode* root) 
        {
            vector<int> values; 

            if (!root) return values;

            vector<TreeNode*> stackTree = {root};
            vector<TreeNode*> visitedNodes;

            while (!stackTree.empty())
            {
                root = stackTree.back();

                if (root->left &&
                find(visitedNodes.begin(), visitedNodes.end(), root->left) == visitedNodes.end())
                {
                    stackTree.push_back(root->left);
                    visitedNodes.push_back(root->left);
                }
                else if (root->right &&
                find(visitedNodes.begin(), visitedNodes.end(), root->right) == visitedNodes.end())
                {
                    stackTree.push_back(root->right);
                    visitedNodes.push_back(root->right);
                }
                else
                {
                    values.push_back(root->val);
                    stackTree.pop_back();
                }
            }    

            return values;
        }
};

File: leetcode/100-999/145. Binary Tree Postorder Traversal/solution_3.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number[]}
 */
 
var postorderTraversal = function(root) 
{
    const values = []

    if (!root) return values

    const stackTree = [root]
    const visitedNodes = []

    while (stackTree.length > 0)
    {
        const root = stackTree[stackTree.length - 1]

        if (root.left && !visitedNodes.includes(root.left))
        {
            stackTree.push(root.left)
            visitedNodes.push(root.left)
        }
        else if (root.right && !visitedNodes.includes(root.right))
        {
            stackTree.push(root.right)
            visitedNodes.push(root.right)
        }
        else
        {
            values.push(root.val)
            stackTree.pop()
        }
    }     

    return values
};